File: resources/[vrp]/vrp-lib/server/sv_rpc.js
const Promises = new Map(), Resource = GetCurrentResourceName();

let CallIdentifier = 0;
function TriggerNetEvent(pName, pSource, ...params) {
    const payload = msgpack_pack(params);
    payload.length < 5000 ? TriggerClientEventInternal(pName, pSource, payload, payload.length) : TriggerLatentClientEventInternal(pName, pSource, payload, payload.length, 128000);
}

function ParamPacker(...params) {
    const pack = [];
    for (let i = 0; i < params.length; i += 1) {
        pack[i] = { 'param': params[i] };
    }
    return pack;
}

function UnPacker(params) {
    const packs = [];
    return params.forEach(pEach => packs.push(pEach.param)), packs;
}

async function rpcRegister(pName, pFunction) {
    onNet('rpc:request:' + pName, async function (origin, callID, params, packaged) {
        const src = source
        let pCallback;
        packaged && (params = UnPacker(params));
        try {
            pCallback = await pFunction(src, ...params);
        } catch (_0x12cd75) {
            emit('rpc:client:error', Resource, origin, pName, _0x12cd75.message);
        }
        if (typeof pCallback === 'undefined') {
            pCallback = [];
        } else {
            packaged && (pCallback = ParamPacker(pCallback));
        }
        TriggerNetEvent('rpc:response', src, origin, callID, pCallback, packaged);
    });
}

async function rpcExecute(pName, pSource, ...params) {
    const callID = CallIdentifier;
    CallIdentifier++;
    const _0x57c5fb = new Promise((resolve, reject) => {
        Promises.set(callID, {
            'resolve': resolve,
            'reject': reject
        });
    });
    TriggerNetEvent('rpc:request:' + pName, pSource, Resource, callID, params, false);
    const _0x5de8f3 = new Promise(resolve => {
        setTimeout(() => {
            if (Promises.has(callID)) {
                Promises.delete(callID);
            }
            return resolve(false);
        }, 31000);
    });
    return await Promise.race([
        _0x57c5fb,
        _0x5de8f3
    ]);
}

const RPC = {
    'register': rpcRegister,
    'execute': rpcExecute
};

onNet('rpc:response', (origin, callID, params, packaged) => {
    Resource === origin && Promises.has(callID) && (packaged && (params = UnPacker(params)), Promises.get(callID).resolve(params), Promises.delete(callID));
});

File: resources/[vrp]/vrp-lib/server/sv_rpc.lua
local Resource, Promises, CallIdentifier = GetCurrentResourceName(), {}, 0

RPC = {}

function ClearPromise(callID)
    Citizen.SetTimeout(5000, function()
        Promises[callID] = nil
    end)
end

function ParamPacker(...)
    local params, pack = {...} , {}

    for i = 1, 15, 1 do
        pack[i] = {param = params[i]}
    end

    return pack
end

function ParamUnpacker(params, index)
    local idx = index or 1

    if idx <= #params then
        return params[idx]["param"], ParamUnpacker(params, idx + 1)
    end
end

function UnPacker(params, index)
    local idx = index or 1

    if idx <= 15 then
        return params[idx], UnPacker(params, idx + 1)
    end
end

function TriggerNetEvent(pEvent, pSrc, ...)
    local payload = msgpack.pack({...})

    if payload:len() < 5000 then
        TriggerClientEventInternal(pEvent, pSrc, payload, payload:len())
    else
        TriggerLatentClientEventInternal(pEvent, pSrc, payload, payload:len(), 128000)
    end
end

------------------------------------------------------------------
--                  (Trigger Server Calls)
------------------------------------------------------------------

function RPC.execute(name, pSource, ...)
    local callID, solved = CallIdentifier, false
    CallIdentifier = CallIdentifier + 1

    Promises[callID] = promise:new()

    TriggerNetEvent("rpc:request:" .. name, pSource, Resource, callID, ParamPacker(...), true)

    Citizen.SetTimeout(20000, function()
        if not solved then
            Promises[callID]:resolve({nil})
            TriggerEvent("rpc:server:timeout", Resource, name)
        end
    end)

    local response = Citizen.Await(Promises[callID])

    solved = true

    ClearPromise(callID)

    return ParamUnpacker(response)
end

RPC.executeLatent = RPC.execute

RegisterNetEvent("rpc:response")
AddEventHandler("rpc:response", function(origin, callID, response)
    if Resource == origin and Promises[callID] then
        Promises[callID]:resolve(response)
    end
end)

------------------------------------------------------------------
--                  (Receive Remote Calls)
------------------------------------------------------------------

function RPC.register(name, func)
    RegisterNetEvent("rpc:request:" .. name, function(origin, callID, params)
        local src = source
        local response

        local success, error = pcall(function()
            response = ParamPacker(func(src, ParamUnpacker(params)))
        end)

        if not success then
            TriggerEvent("rpc:client:error", Resource, origin, name, error)
        end

        if response == nil then
            response = {}
        end

        TriggerNetEvent("rpc:response", src, origin, callID, response, true)
    end)
end

function RPC.remove(name)
end